File: Dynamic-Programming/188-H-Best-Time-to-Buy-and-Sell-Stock-IV/best_time_to_buy_iv.cpp
class Solution {
public:
    int maxProfit(int k, vector<int>& prices) {
        int n = prices.size();
        if (n <= 1)
            return 0;
        
        vector<vector<int>> OPT(k+1, vector<int>(n, 0));
        for (int i=1; i<=k; i++) {
                int maxDiff = -prices[0];
            for (int j=1; j<n; j++) {
                OPT[i][j] = max(OPT[i][j-1], prices[j]+maxDiff);
                maxDiff = max(maxDiff, OPT[i-1][j]-prices[j]);
            }
        }
        return OPT[k][n-1];
    }
};

File: Dynamic-Programming/188-H-Best-Time-to-Buy-and-Sell-Stock-IV/best_time_to_buy_iv.py
class Solution:
    def maxProfit(self, k: int, prices: list[int]) -> int:
        n = len(prices)
        if n <= 1:
            return 0
        
        OPT = [[0]*n for _ in range(k+1)]
        for i in range(1, k+1):
            maxDifference = -prices[0]
            for j in range(1, n):
                OPT[i][j] = max(OPT[i][j-1], maxDifference+prices[j])
                maxDifference = max(maxDifference, OPT[i-1][j] - prices[j])
        return OPT[k][n-1]